### PR TITLE
Improve the multi party pattern

### PR DESCRIPTION
**Pull request by bethaitman**
I'd be very interested in comments here - this was meant to make it easier for beginners to follow, which might mean it's become worse in some other ways.

Summary of changes:

- Removed the `toSign` list as a parameter, because you can infer it from other details
- Experimenting with clearer names
- More code comments, including a clarifying note about the contents of the end contract
- Added ensure conditions
- Added extra tests in the scenario

### CODE REVIEW

File: docs/source/daml/patterns/daml/MultiplePartyAgreement.daml [hunk 1]
```diff
@@ -11,60 +11,82 @@ import Utilities
...
       controller signer
         do
+          -- Check the controller is in the toSign list, and if they are, sign the Pending contract
           assert (signer `elem` toSign)
-          create this with signed = signer :: signed, toSign = delete signer toSign
+          create this with alreadySigned = signer :: alreadySigned
     -- end snippet: first half pending template
     -- start snippet: second half pending template
-    choice Finalize : ContractId Agreement with
+    choice Finalize : ContractId ContractPlaceholder with
```
Comment (recurse) [inline docs/source/daml/patterns/daml/MultiplePartyAgreement.daml]: I think it is important to show how to auto-finalize the contract, even if we show this one first as the basic manipulation of `Either` and `Optional` within `Update` is a basic DAML skill.

File: docs/source/daml/patterns/daml/MultiplePartyAgreement.daml [hunk 2]
```diff
@@ -11,60 +11,82 @@ import Utilities
...
+    signatory alreadySigned
+    observer finalContract.signatories
+    ensure
+      -- Can't have duplicate signatories
+      unique alreadySigned
+
+    -- The parties who need to sign is the finalContract.signatories with alreadySigned filtered out
+    let toSign = filter (`notElem` alreadySigned) finalContract.signatories
 
     choice Sign : ContractId Pending with
```
Comment (recurse) [inline docs/source/daml/patterns/daml/MultiplePartyAgreement.daml]: Flexible controller choice is a nice change.

File: docs/source/daml/patterns/daml/MultiplePartyAgreement.daml (hunk 1 above)
Comment (bethaitman) [reply docs/source/daml/patterns/daml/MultiplePartyAgreement.daml]: I'd like to do this, but within a separate pattern.

File: docs/source/daml/patterns/daml/MultiplePartyAgreement.daml (hunk 2 above)
Comment (bethaitman) [reply docs/source/daml/patterns/daml/MultiplePartyAgreement.daml]: I agree!

**Pull request merged**